pkg/host-disk/host-disk.go

```diff
@@ -85,19 +91,58 @@ func getPVCDiskImgPath(volumeName string) string {
 	return path.Join(pvcBaseDir, volumeName, "disk.img")
 }
 
-func CreateHostDisks(vmi *v1.VirtualMachineInstance) error {
+type DiskImgCreator struct {
+	dirBytesAvailableFunc  func(path string) (uint64, error)
+	notifier               k8sNotifier
+	lessPVCSpaceToleration int
+}
+
+type k8sNotifier interface {
+	SendK8sEvent(vmi *v1.VirtualMachineInstance, severity string, reason string, message string) error
+}
+
+func NewHostDiskCreator(notifier k8sNotifier, lessPVCSpaceToleration int) DiskImgCreator {
+	return DiskImgCreator{
+		dirBytesAvailableFunc:  dirBytesAvailable,
+		notifier:               notifier,
+		lessPVCSpaceToleration: lessPVCSpaceToleration,
+	}
+}
+
+func (hdc *DiskImgCreator) setlessPVCSpaceToleration(toleration int) {
+	hdc.lessPVCSpaceToleration = toleration
+}
+
+func (hdc DiskImgCreator) Create(vmi *v1.VirtualMachineInstance) error {
 	for _, volume := range vmi.Spec.Volumes {
 		if hostDisk := volume.VolumeSource.HostDisk; hostDisk != nil && hostDisk.Type == v1.HostDiskExistsOrCreate && hostDisk.Path != "" {
 			if _, err := os.Stat(hostDisk.Path); os.IsNotExist(err) {
-				availableSpace, err := dirBytesAvailable(path.Dir(hostDisk.Path))
+				availableSize, err := hdc.dirBytesAvailableFunc(path.Dir(hostDisk.Path))
 				if err != nil {
 					return err
 				}
-				size, _ := hostDisk.Capacity.AsInt64()
-				if uint64(size) > availableSpace {
-					return fmt.Errorf("Unable to create %s, not enough space, demanded size %d B is bigger than available space %d B", hostDisk.Path, uint64(size), availableSpace)
+				requestedSize, _ := hostDisk.Capacity.AsInt64()
+				diskSize := requestedSize
+				if uint64(requestedSize) > availableSize {
+					// Some storage provisioners provision less space than requested, due to filesystem overhead etc.
+					// We tolerate some difference in requested and available capacity up to some degree.
+					// This can be configured with the "pvc-tolerate-less-space-up-to-percent" parameter in the kubevirt-config ConfigMap.
+					// It is provided as argument to virt-launcher.
+					toleratedSize := requestedSize * (100 - int64(hdc.lessPVCSpaceToleration)) / 100
+					if uint64(toleratedSize) > availableSize {
+						return fmt.Errorf("unable to create %s, not enough space, demanded size %d B is bigger than available space %d B, also after taking %v %% toleration into account",
+							hostDisk.Path, uint64(requestedSize), availableSize, hdc.lessPVCSpaceToleration)
+					}
+					diskSize = int64(availableSize)
+
+					msg := fmt.Sprintf("PV size too small: expected %v B, found %v B. Using it anyway, it is within %v %% toleration", requestedSize, availableSize, hdc.lessPVCSpaceToleration)
```

Comment:
Something is not  right here. I see

```
kubevirt-test-default   5s        5s        3         testvmikqcx6zjv8k9kfdhq4p96k4whlt7mbbjlp6g9lpz9xb6mjsdt.15696bb0f4c21ea2   VirtualMachineInstance             Warning   SyncFailed   virt-handler, node01   server error. command Launcher.Sync failed: unable to create /tmp/kubevirt/krzs9chpww/mount/disk.img, not enough space, demanded size 39705950400 B is bigger than available space 33881468928 B, also after taking 10 %!t(MISSING)oleration into account
```
in the event log.

Reply:
the same formatting is used for the ToleratedSmallPV event, there it works. I guess in case of an error, the formatted error message ("...taking 10 % toleration...") is used again for formatting somewhere else, and there the single % fails... will try find that place